pkg/agent/openflow/pipeline.go

```diff
@@ -41,38 +41,74 @@ import (
-	ServiceHairpinTable          = binding.NewOFTable(23, "ServiceHairpin")
-	ServiceConntrackTable        = binding.NewOFTable(24, "ServiceConntrack") // serviceConntrackTable use a new ct_zone to transform SNAT'd connections.
-	ConntrackTable               = binding.NewOFTable(30, "ConntrackZone")
-	ConntrackStateTable          = binding.NewOFTable(31, "ConntrackState")
-	ServiceClassifierTable       = binding.NewOFTable(35, "ServiceClassifier")
-	SessionAffinityTable         = binding.NewOFTable(40, "SessionAffinity")
-	DNATTable                    = binding.NewOFTable(40, "DNAT(SessionAffinity)")
-	ServiceLBTable               = binding.NewOFTable(41, "ServiceLB")
-	EndpointDNATTable            = binding.NewOFTable(42, "EndpointDNAT")
-	AntreaPolicyEgressRuleTable  = binding.NewOFTable(45, "AntreaPolicyEgressRule")
-	DefaultTierEgressRuleTable   = binding.NewOFTable(49, "DefaultTierEgressRule")
-	EgressRuleTable              = binding.NewOFTable(50, "EgressRule")
-	EgressDefaultTable           = binding.NewOFTable(60, "EgressDefaultRule")
-	EgressMetricTable            = binding.NewOFTable(61, "EgressMetric")
-	L3ForwardingTable            = binding.NewOFTable(70, "L3Forwarding")
-	SNATTable                    = binding.NewOFTable(71, "SNAT")
-	L3DecTTLTable                = binding.NewOFTable(72, "IPTTLDec")
-	L2ForwardingCalcTable        = binding.NewOFTable(80, "L2Forwarding")
-	AntreaPolicyIngressRuleTable = binding.NewOFTable(85, "AntreaPolicyIngressRule")
-	DefaultTierIngressRuleTable  = binding.NewOFTable(89, "DefaultTierIngressRule")
-	IngressRuleTable             = binding.NewOFTable(90, "IngressRule")
-	IngressDefaultTable          = binding.NewOFTable(100, "IngressDefaultRule")
-	IngressMetricTable           = binding.NewOFTable(101, "IngressMetric")
-	ConntrackCommitTable         = binding.NewOFTable(105, "ConntrackCommit")
-	ServiceConntrackCommitTable  = binding.NewOFTable(106, "ServiceConntrackCommit")
-	HairpinSNATTable             = binding.NewOFTable(108, "HairpinSNAT")
-	L2ForwardingOutTable         = binding.NewOFTable(110, "Output")
+	// PipelineClassifierTable is table 0. Packets are forwarded to different pipelines in this table.
+	PipelineClassifierTable = newTable("PipelineClassifier", 0xff, 0)
+
+	// OVS pipeline for ARP is used to process ARP packets.
+	// Tables in ValidationStage:
+	ARPSpoofGuardTable = newTable("ARPSpoofGuard", 0x7f, binding.ValidationStage)
+
+	// Tables in OutputStage:
+	ARPResponderTable = newTable("ARPResponder", 0x7f, binding.OutputStage)
+
+	// OVS pipeline for IP is used to process IPv4/IPv6 packets.
+	// Tables in ClassifierStage:
+	ClassifierTable = newTable("Classifier", 0x7f, binding.ClassifierStage)
+
+	// Tables in ValidationStage:
+	SpoofGuardTable   = newTable("SpoofGuard", 0x7f, binding.ValidationStage)
+	IPv6Table         = newTable("IPv6", 0x6f, binding.ValidationStage)
+	IPClassifierTable = newTable("IPClassifier", 0x5f, binding.ValidationStage)
```

Comment:
Following my comment in https://github.com/antrea-io/antrea/pull/3058#discussion_r814951542.
I had a talk with @hongliangl to understand the necessity of "stage" and two points I got were:
1. grouping similar tables together
2. determing the order of tables by leveraging the static order of stages and the dynamic priority of tables in a stage.

I agree with the purpose but I think it could be achieved by some simpler approach, introducing less concepts, structs, code:
Instead of introducing stage explicitly and having a priority value for each table, it could just list the desired order of all possible tables in a slice, which should be more declarative than using the static order of stages and the dynamic priority. The stage concept can just exist in the comment, used to help understand the logic of pipeline. And it doesn't have to to named as a stage if it's not suitable.
```
var Pipeline = []*Table{
	...
	// PreRoutingStage, which takes care of DNAT.
	// When proxy is enabled
	PreRoutingClassifierTable,
	NodePortMarkTable,
	SessionAffinityTable,
	ServiceLBTable,
	EndpointDNATTable,
	// When proxy is disabled
	DNATTable,

	// EgressSecurity tables
	AntreaPolicyEgressRuleTable,
	EgressRuleTable,
	EgressDefaultTable,
	EgressMetricTable,

	// RoutingStage
	L3ForwardingTable,
	ServiceHairpinMarkTable,
	L3DecTTLTable,

	// PostRoutingStage, which takes care of SNAT.
	SNATTable,
	SNATConntrackCommitTable,
	...
}
```
Then when declaring the required tables for a feature, it could just use a slice/set of tables, instead of the current 3-dimensional data: pipeline -> stage -> table. It looks to me at least stage -> table is redundant with the definition of table: table is defined with stage specified, why it needs to declare the stage again when declaring the required tables, and what if it specifies a mismatched stage? The updated code example could look like:
```
func (f *featurePodConnectivity) getRequiredTables() []*Table {
	requriedTables := []*Table{
		ClassifierTable,
		SpoofGuardTable,
		ConntrackTable,
		ConntrackStateTable,
		...
	}
	if f.enableMulticast {
		requriedTables = append(requriedTables, IPClassifierTable)
	}
	...
	return requiredTables
}
```
For the purpose that we want to make it easy and reliable to jump to specific stage without concerning the change of first table of a stage, it could be achieved by declaring an alias for required stage, such as
```
var (
	EntrancePreRoutingStage     = PreRoutingClassifierTable
	EntranceEgressPolicy        = AntreaPolicyEgressRuleTable
	EntranceRoutingStage        = L3ForwardingTable
	EntrancePostRoutingStage    = SNATTable
)

/*
    // For most cases, packets should go to next table declared by the pipeline.
	L3ForwardingTable.ofTable.BuildFlow(priorityMiss).
		Cookie(cookieID).
		Action().LoadRegMark(EgressRegMark).
		Action().LoadRegMark(ToGatewayRegMark).
		Action().GoToNextTable().
		Done(),

    // For specific cases, packets could jump to specific stage or table when necessary. For example,
    // - when there is an obvious performance gain.
    // - when specific kinds of traffic shouldn't go through the whole pipeline, like ARP, multicast.
	L3ForwardingTable.ofTable.BuildFlow(priorityMiss).
		Cookie(cookieID).
		Action().LoadRegMark(EgressRegMark).
		Action().LoadRegMark(ToGatewayRegMark).
		Action().GotoTable(EntrancePostRoutingStage).
		Done(),
*/
```

Another reason for not introducing a real type "stage" is, even for now it doesn't seem suiting the requirement well. The code seems to create some stage names to fit in the rule that a packet should always to go to the first table of a "stage", for example, IngressSecurityStage, ConntrackStage, ConntrackStateStage. I'm not sure if most people can imagine what's the desired order of ConntrackStage and ConntrackStateStage because they are not really "stage" but are promoted from table to stage because of the aforementioned rules. And the current interfaces cannot enforce the rules strictly, still up to developers' subjectivity.
@jianjuns @antoninbas @wenyingd @hongliangl